sql/core/src/main/scala/org/apache/spark/sql/Arrow.scala

```diff
@@ -123,67 +86,21 @@ object Arrow {
       rows: Array[InternalRow],
       ordinal: Int,
       field: StructField,
-      allocator: RootAllocator): (Array[ArrowBuf], Array[ArrowFieldNode]) = {
+      allocator: RootAllocator): (Array[ArrowFieldNode], Array[ArrowBuf]) = {
     val numOfRows = rows.length
-
-    field.dataType match {
-      case IntegerType | LongType | DoubleType | FloatType | BooleanType | ByteType =>
-        val validityVector = new BitVector("validity", allocator)
-        val validityMutator = validityVector.getMutator
-        validityVector.allocateNew(numOfRows)
-        validityMutator.setValueCount(numOfRows)
-
-        val buf = allocator.buffer(numOfRows * field.dataType.defaultSize)
-        val typeFunc = getTypeFuncs(field.dataType)
-        var nullCount = 0
-        var index = 0
-        while (index < rows.length) {
-          val row = rows(index)
-          if (row.isNullAt(ordinal)) {
-            nullCount += 1
-            validityMutator.set(index, 0)
-            typeFunc.fill(buf)
-          } else {
-            validityMutator.set(index, 1)
-            typeFunc.write(row, ordinal, buf)
-          }
-          index += 1
-        }
-
-        val fieldNode = new ArrowFieldNode(numOfRows, nullCount)
-
-        (Array(validityVector.getBuffer, buf), Array(fieldNode))
-
-      case StringType =>
-        val validityVector = new BitVector("validity", allocator)
-        val validityMutator = validityVector.getMutator()
-        validityVector.allocateNew(numOfRows)
-        validityMutator.setValueCount(numOfRows)
-
-        val bufOffset = allocator.buffer((numOfRows + 1) * IntegerType.defaultSize)
-        var bytesCount = 0
-        bufOffset.writeInt(bytesCount)
-        val bufValues = allocator.buffer(1024)
-        var nullCount = 0
-        rows.zipWithIndex.foreach { case (row, index) =>
-          if (row.isNullAt(ordinal)) {
-            nullCount += 1
-            validityMutator.set(index, 0)
-            bufOffset.writeInt(bytesCount)
-          } else {
-            validityMutator.set(index, 1)
-            val bytes = row.getUTF8String(ordinal).getBytes
-            bytesCount += bytes.length
-            bufOffset.writeInt(bytesCount)
-            bufValues.writeBytes(bytes)
-          }
-        }
-
-        val fieldNode = new ArrowFieldNode(numOfRows, nullCount)
-
-        (Array(validityVector.getBuffer, bufOffset, bufValues),
-            Array(fieldNode))
+    val columnWriter = schemaToColumnWriter(allocator, field.dataType)
+    columnWriter.init(numOfRows)
+    var index = 0
+
+    while(index < numOfRows) {
+      val data = rows(index).get(ordinal, field.dataType)
+      columnWriter.write(data)
```

Comment:
I think it might be a little clearer to check for null here, like before and just define `writeData` and `writeNull` in the `ColumnWriter` interface.  Is that do-able?